pkg/ingester/limiting.go

```diff
@@ -0,0 +1,239 @@
+	var memUtil uint64
+	if _, err := fmt.Fscanf(memR, "%d", &memUtil); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", memPath)
+	}
+
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
+	const cpuPath = "/sys/fs/cgroup/cpu.stat"
+	cpuR, err := readFileNoStat(cpuPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	var name string
+	var cpuTime uint64
+	for {
+		if _, err := fmt.Fscanf(cpuR, "%s %d\n", &name, &cpuTime); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPath)
+		}
+
+		if name == "usage_usec" {
+			// This is CPU time in microseconds
+			break
+		}
+	}
+	if name != "usage_usec" {
+		return 0, 0, fmt.Errorf("failed scanning %s", cpuPath)
+	}
+
+	return float64(cpuTime) / 1e06, memUtil, nil
+}
+
+func scanCgroupV1() (float64, uint64, error) {
+	const memPath = "/sys/fs/cgroup/memory/memory.stat"
+	memR, err := readFileNoStat(memPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	// Summing RSS and cache for memory usage
+	// For reference, see section 5.5 in https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
+	var memUtil uint64
+	var numMemStats int
+	for {
+		var name string
+		var val uint64
+		if _, err := fmt.Fscanf(memR, "%s %d\n", &name, &val); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", memPath)
+		}
+
+		if name == "rss" || name == "cache" {
+			memUtil += val
+			numMemStats++
+			if numMemStats == 2 {
+				break
+			}
+		}
+	}
+	if numMemStats != 2 {
+		return 0, 0, fmt.Errorf("failed scanning %s", memPath)
+	}
+
+	// For reference, see https://www.kernel.org/doc/Documentation/cgroup-v1/cpuacct.txt
+	const cpuPath = "/sys/fs/cgroup/cpu/cpuacct.usage"
+	cpuR, err := readFileNoStat(cpuPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	// CPU time in nanoseconds
+	var cpuTime uint64
+	if _, err := fmt.Fscanf(cpuR, "%d", &cpuTime); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPath)
+	}
+
+	return float64(cpuTime) / 1e09, memUtil, nil
+}
+
+func scanProcFS() (float64, uint64, error) {
+	p, err := procfs.Self()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process info")
+	}
+	ps, err := p.Stat()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process stats")
+	}
+
+	return ps.CPUTime(), uint64(ps.ResidentMemory()), nil
+}
+
+func (i *Ingester) updateResourceUtilization() {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return
+	}
+
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+
+	var method string
+	cpuTime, memUtil, err := scanCgroupV2()
+	if err != nil && !os.IsNotExist(err) {
+		level.Warn(i.logger).Log("msg", "failed to get CPU and memory stats from cgroup v2", "err", err.Error())
+		return
+	}
+	if err == nil {
+		method = "cgroup v2"
```

Comment:
Design wise, what do you think if we check which one is supported only once at the beginning and then we always use it? I would clean up a bit the code defining an interface with methods to read the CPU and memory utilization, then 3 implementations: cgroup v2, cgroup v1 and proc (plus a no-op for the case the system is not supported). At the ingester startup, we look which implementation is supported and we use the right implementation.

Reply:
Last but not the least, all of this is unrelated from the ingester, so you could consider moving it to `pkg/util/...`.

Reply:
> Design wise, what do you think if we check which one is supported only once at the beginning and then we always use it? I would clean up a bit the code defining an interface with methods to read the CPU and memory utilization, then 3 implementations: cgroup v2, cgroup v1 and proc (plus a no-op for the case the system is not supported). At the ingester startup, we look which implementation is supported and we use the right implementation.

I've done the above and moved `UtilizationBasedLimiter` to pkg/util/limiter.